packages/commons/src/router/errorHandlers.ts

```diff
@@ -0,0 +1,74 @@
+import { constants } from "http2";
+import express, { Response, NextFunction } from "express";
+import {
+  badRequestError,
+  genericError,
+  makeApiProblemBuilder,
+  parseErrorMessage,
+} from "pagopa-interop-models";
+import { z } from "zod";
+import { fromZodIssue } from "zod-validation-error";
+import { WithZodiosContext } from "@zodios/express";
+import { ExpressContext, fromAppContext } from "../context/context.js";
+
+const makeApiProblem = makeApiProblemBuilder({});
+
+export function zodiosValidationErrorToApiProblem(
+  zodError: {
+    context: string;
+    error: z.ZodIssue[];
+  },
+  req: WithZodiosContext<express.Request, ExpressContext>,
+  res: Response,
+  _next: NextFunction
+): Response {
+  const ctx = fromAppContext(req.ctx);
+  const detail = `Incorrect value for ${zodError.context}`;
+  const errors = zodError.error.map((e) => fromZodIssue(e));
+
+  return res
+    .status(constants.HTTP_STATUS_BAD_REQUEST)
+    .send(
+      makeApiProblem(
+        badRequestError(detail, errors),
+        () => constants.HTTP_STATUS_BAD_REQUEST,
+        ctx
+      )
+    );
+}
+
+export function errorsToApiProblemsMiddleware(
+  error: unknown,
+  req: WithZodiosContext<express.Request, ExpressContext>,
+  res: Response,
+  next: NextFunction
+): Response | void {
+  if (res.headersSent) {
```

Comment:
I'm not aware of the mechanism, is this the discriminator to tell if a different error has been already handled?
For instance, if the endpoint wants to return a 404 based on the business logic, and since this middleware is the last one to be executed, which is the check that avoids landing on [this](https://github.com/pagopa/interop-be-monorepo/pull/2176/files#diff-7f9d707b280ed88cb36984728e84eae9e20f7f5bf8d6dcd1a0344748c02e28edR66) internal error?

Reply:
@galales in our routers we always catch errors and send them in response with `return res.status(errorRes.status).send(errorRes)` - this is what triggers `res.headersSent` to be true.
This function replaces the express default error handler to handle any error that reaches this last middleware: see docs here https://expressjs.com/en/guide/error-handling.html#the-default-error-handler

(Note that since we do not call `next` after we send responses in routers, we should not reach this default error handler anyways whenever we send another error before - but adding the check is safer and suggested by the docs)